### PR TITLE
Add PRAuC

### PR DESCRIPTION
**Pull request by Dref360**
## Summary:
Add a Precision-Recall Area under the curve.

### Features:
* New metrics suitable for imbalanced datasets.

## Checklist:

* [x] Your code is documented (To validate this, add your module to `tests/documentation_test.py`).
* [x] Your code is tested with unit tests.
* [x] You moved your Issue to the PR state.

### CODE REVIEW

File: src/baal/utils/metrics.py [hunk 1]
```diff
@@ -1,10 +1,28 @@
...
+    Args:
+        probabilities (tensor): [batch_size, num_classes, ...]
+
+    Returns:
+        Same as probabilities.
+    """
+    bounded = torch.min(probabilities) < 0 or torch.max(probabilities) > 1.0
+    if bounded or not torch.allclose(probabilities.sum(1), torch.ones_like(probabilities)[..., 0]):
+        probabilities = torch.softmax(probabilities, 1)
+    return probabilities
```
Comment (parmidaatg) [inline src/baal/utils/metrics.py]: we already have this in heuristics, should we put it in utils for the usage of both?

File: src/baal/utils/metrics.py [hunk 2]
```diff
@@ -98,9 +102,24 @@ def __init__(self, n_bins=10, **kwargs):
...
+        """
+        Updating the true positive (tp) and number of samples in each bin.
+
+        Args:
+            output (tensor): logits or predictions of model
+            target (tensor): labels
+        """
+        output = output.detach().cpu().numpy()
+        target = target.detach().cpu().numpy()
+        output = to_prob(output)
```
Comment (IssamLaradji) [inline src/baal/utils/metrics.py]: is there a reason why softmax is not called explicitly and replaced by to_prob instead?

File: src/baal/utils/metrics.py [hunk 3]
```diff
@@ -307,3 +326,73 @@ def value(self):
...
+        """
+        Update the confusion matrice according to output and target.
+
+        Args:
+            output (tensor): predictions of model
+            target (tensor): labels
+        """
+        output = output.detach().cpu().numpy()
+        target = target.detach().cpu().numpy()
+        output = to_prob(output)
```
Comment (IssamLaradji) [inline src/baal/utils/metrics.py]: this is kind of risky if the output is already softmaxed. Wouldn't it be better if you do an assertion instead? For instance, you make this function expect probabilities and and assert that the probs add up to 1. This reduces the amount possible silent bugs.

Comment (Dref360) [reply src/baal/utils/metrics.py]: If the output sums to 1 and they are between 0 and 1 we don't softmax it.

Comment (IssamLaradji) [reply src/baal/utils/metrics.py]: ah oki, makes sense!

File: src/baal/utils/metrics.py (hunk 2 above)
Comment (orangetoaster) [reply src/baal/utils/metrics.py]: It looks like in the wrapper case, to_prop was extracted for clarity. Then it's best to use it consistently so that if it changes in the to_prob method, it'll take effect everywhere. In this case we only want to take the softmax if there are values outside [0,1] or if they don't sum to 1.

File: src/baal/utils/metrics.py [hunk 4]
```diff
@@ -307,3 +334,73 @@ def value(self):
...
+                report = self._make_report(output[:, cls, ...], target_cls, th)
+                self._data[cls][th].fp += report.fp
+                self._data[cls][th].tp += report.tp
+                self._data[cls][th].fn += report.fn
+
+    def _make_report(self, output, target, threshold):
+        output = (output > threshold).astype(np.int8)
+        output = output.reshape([-1])
+        target = target.reshape([-1])
+        _, fp, fn, tp = confusion_matrix(target, output, labels=[0, 1]).ravel()
```
Comment (orangetoaster) [inline src/baal/utils/metrics.py]: why not use `.reshape(-1, order='C')` ?
